File: app/src/core/domain/identifiers.ts
import type { CountryProfile } from '../../shared/types'
import { activeProfile } from './active-profile'

/**
 * FHIR and export identifiers, built from the active country profile.
 *
 * These strings are wire-visible: every FHIR resource and aggregate payload this app
 * emits carries them, and downstream consumers match on them. The India profile
 * reproduces the previously hardcoded values exactly, so existing deployments emit
 * byte-identical output.
 *
 * Mirrors server/amrit_central_server/central/identifiers.py; both read the same profile.
 */

const DEFAULT_BASE_URI = 'https://amrit.invalid'
const DEFAULT_URN_PREFIX = 'urn:amrit'

const namespaceOf = (profile: CountryProfile): { base_uri?: string; urn_prefix?: string } =>
  profile.identifier_namespace ?? {}

export function baseUri(profile: CountryProfile = activeProfile()): string {
  return (namespaceOf(profile).base_uri || DEFAULT_BASE_URI).replace(/\/+$/, '')
}

export function urnPrefix(profile: CountryProfile = activeProfile()): string {
  return (namespaceOf(profile).urn_prefix || DEFAULT_URN_PREFIX).replace(/:+$/, '')
}

export function uri(segments: string[], profile: CountryProfile = activeProfile()): string {
  const path = segments.map((segment) => String(segment).replace(/^\/+|\/+$/g, '')).filter(Boolean).join('/')
  return path ? `${baseUri(profile)}/${path}` : baseUri(profile)
}

export function urn(segments: string[], profile: CountryProfile = activeProfile()): string {
  const tail = segments.map((segment) => String(segment).replace(/^:+|:+$/g, '')).filter(Boolean).join(':')
  return tail ? `${urnPrefix(profile)}:${tail}` : urnPrefix(profile)
}

/** Named helpers, so each identifier's shape lives in exactly one place. */
export const aggregateIdentifierSystem = (profile?: CountryProfile): string => urn(['aggregate'], profile ?? activeProfile())
export const laboratoryIdentifierSystem = (profile?: CountryProfile): string => urn(['laboratory'], profile ?? activeProfile())
export const resistanceMeasureUrn = (antibioticCode: string, profile?: CountryProfile): string =>
  urn(['measure', 'resistance-rate', antibioticCode], profile ?? activeProfile())
export const aggregateCodeSystem = (profile?: CountryProfile): string =>
  uri(['CodeSystem', 'aggregate'], profile ?? activeProfile())

/**
 * The canonical URL of an AMRIT profile, for `meta.profile` and for the Implementation Guide.
 *
 * Phase 25. It hangs off the deployment's own base URI rather than a domain this project
 * invented, for the same reason every other identifier does: the software is country-neutral,
 * and a canonical pointing at one country's institute would be wrong everywhere else. A
 * deployment that publishes an IG sets `identifier_namespace.base_uri` and its resources
 * immediately claim conformance to profiles at that address; one that does not gets
 * `https://amrit.invalid`, which is a reserved TLD and therefore visibly not a claim.
 */
export const profileCanonical = (name: string, profile?: CountryProfile): string =>
  uri(['StructureDefinition', name], profile ?? activeProfile())

File: server/amrit_central_server/central/identifiers.py
"""FHIR and schema identifiers, built from the active country profile.

These strings are wire-visible: every FHIR resource this server emits carries them, and
downstream consumers match on them. Changing a namespace after data has been exported
splits identifier continuity, which is why the profile treats it as a deliberate setting
rather than something derived.

The India profile reproduces the previously hardcoded values exactly, so existing
deployments emit byte-identical output.
"""

from __future__ import annotations

DEFAULT_BASE_URI = "https://amrit.invalid"
DEFAULT_URN_PREFIX = "urn:amrit"


def _namespace(profile: dict | None) -> dict:
    return (profile or {}).get("identifier_namespace") or {}


def base_uri(profile: dict | None = None) -> str:
    if profile is None:
        profile = _active_profile()
    return str(_namespace(profile).get("base_uri") or DEFAULT_BASE_URI).rstrip("/")


def urn_prefix(profile: dict | None = None) -> str:
    if profile is None:
        profile = _active_profile()
    return str(_namespace(profile).get("urn_prefix") or DEFAULT_URN_PREFIX).rstrip(":")


def _active_profile() -> dict | None:
    from .country_profile import ProfileError, get_profile

    try:
        return get_profile()
    except ProfileError:
        # A misconfigured profile must not stop an export; the reserved .invalid default
        # makes the unconfigured state obvious rather than silently borrowing a namespace.
        return None


def uri(*segments: str, profile: dict | None = None) -> str:
    """Build a canonical URI under the deployment's namespace."""
    path = "/".join(str(segment).strip("/") for segment in segments if str(segment).strip("/"))
    return f"{base_uri(profile)}/{path}" if path else base_uri(profile)


def urn(*segments: str, profile: dict | None = None) -> str:
    """Build a URN under the deployment's prefix, joining segments with colons."""
    tail = ":".join(str(segment).strip(":") for segment in segments if str(segment).strip(":"))
    return f"{urn_prefix(profile)}:{tail}" if tail else urn_prefix(profile)


# Named helpers, so the shape of each identifier lives in one place.
def lab_code_system(profile: dict | None = None) -> str:
    return uri("lab-code", profile=profile)


def measure_url(antibiotic_code: str, profile: dict | None = None) -> str:
    return uri("Measure", f"resistance-rate-{str(antibiotic_code).lower()}", profile=profile)


def proportion_ci_extension(profile: dict | None = None) -> str:
    return uri("StructureDefinition", "proportion-ci", profile=profile)


def bundle_identifier_system(profile: dict | None = None) -> str:
    return uri("bundle-id", profile=profile)


def aggregate_code_system(profile: dict | None = None) -> str:
    return uri("CodeSystem", "aggregate", profile=profile)
